plugins/woocommerce/client/blocks/assets/js/base/context/hooks/cart/use-store-cart.ts

```diff
@@ -93,31 +96,77 @@ const decodeValues = <
 	) as T;
 };
 
+// Normalize address fields to ensure they are always in the same format and update the ref to track the latest value.
+const normalizeAddress = <
+	T extends CartResponseBillingAddress | CartResponseShippingAddress
+>(
+	address: T,
+	addressRef: React.MutableRefObject< T >
+): T => {
+	const normalizedAddress = emptyHiddenAddressFields(
+		decodeValues( address )
+	);
+	if ( ! fastDeepEqual( addressRef.current, normalizedAddress ) ) {
+		addressRef.current = normalizedAddress;
+	}
+	return addressRef.current;
+};
+
+const normalizeCoupons = ( coupons: CartResponseCouponItem[] ) => {
+	return coupons.length > 0
```

Comment:
We could be more defensive here and not just reply on the length check. Same for fees and rated

Reply:
FWIW This is the same code as before just moved into a utility function. I think we can trust the types here; this comes from `cartData` which in terms comes from a hook. I think the there were an unexpected type here we'd get problems in many other places before this.